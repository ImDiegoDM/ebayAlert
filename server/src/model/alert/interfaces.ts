import * as validator from 'validator'
import { ObjError } from '../../objError';

export interface SearchPhrase{
  phrase:string;
  howOften:2|10|30
}

export interface Alert{
  id?:string,
  email:string,
  phrases:SearchPhrase[]
}

export const collection = 'search_phrases';
