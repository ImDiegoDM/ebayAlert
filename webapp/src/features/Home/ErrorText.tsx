import styled from 'styled-components';
export const ErrorText = styled.div`
  margin-top: 10px;
  color: red;
`;
