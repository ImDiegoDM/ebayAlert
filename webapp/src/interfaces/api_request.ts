import { AxiosError } from 'axios';

export interface IApiRequest<T= any, E= any> {
  fetching: boolean;
  fetched: boolean;
  ok: boolean;
  data?: T;
  error?: AxiosError;
}
