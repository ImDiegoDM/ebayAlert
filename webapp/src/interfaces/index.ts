export * from './api_request';
export * from './search_phrase';
