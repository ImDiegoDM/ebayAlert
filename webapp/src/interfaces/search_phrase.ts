export interface ISearchPhrase {
  phrase: string;
  email: string;
  howOften: '2' | '10' | '30';
}
