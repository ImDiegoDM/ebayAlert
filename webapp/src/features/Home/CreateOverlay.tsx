import axios from 'axios';
import React, { useState } from 'react';
import { useDispatch } from 'react-redux';
import { fetchPhrases } from '../../actions';
import { ISearchPhrase } from '../../interfaces';
import { Overlay } from './components/Overlay';
import { SearchPhraseForm } from './components/SearchPhraseForm';
import { ErrorText } from './ErrorText';

interface CreateOverlayProps {
  open: boolean;
  onClose?: () => void;
}

function CreateSearchPhrase(value: ISearchPhrase) {
  return axios.post(process.env.REACT_APP_SERVER_URL + '/search_phrases', value);
}

export function CreateOverlay(props: CreateOverlayProps) {
  const [value, setValue] = useState<ISearchPhrase>({
    email: '',
    howOften: '2',
    phrase: '',
  });

  const [error, setError] = useState('');

  const dispatch = useDispatch();

  function handleSubmit() {
    CreateSearchPhrase(value).then((res) => {
      if (props.onClose) {
        props.onClose();
        setError('');
      }

      dispatch(fetchPhrases());
    }).catch((err) => {
      setError(err.response.data);
    });
  }

  return <Overlay open={props.open} onExit={() => {
    if (props.onClose) {
      props.onClose();
    }
    setError('');
  }}>
    Add a Search Phrase
    {error && <ErrorText>{error}</ErrorText>}
    <SearchPhraseForm value={value} onValueChange={setValue} onSubmit={handleSubmit}/>
  </Overlay>;
}
