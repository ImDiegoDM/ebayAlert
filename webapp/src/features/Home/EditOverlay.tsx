import axios from 'axios';
import React, { useEffect, useState } from 'react';
import { useDispatch } from 'react-redux';
import { fetchPhrases } from '../../actions';
import { ISearchPhrase } from '../../interfaces';
import { Overlay } from './components/Overlay';
import { SearchPhraseForm } from './components/SearchPhraseForm';
import { ErrorText } from './ErrorText';

interface EditOverlayProps {
  open: boolean;
  onClose?: () => void;
  phrase?: ISearchPhrase;
}

function EditSearchPhrase(value: ISearchPhrase) {
  return axios.put(process.env.REACT_APP_SERVER_URL + '/search_phrases/' + value._id, {
    email: value.email,
    howOften: value.howOften,
    phrase: value.phrase,
  });
}

export function EditOverlay(props: EditOverlayProps) {
  const [value, setValue] = useState<ISearchPhrase|undefined>(undefined);
  const [error, setError] = useState('');

  useEffect(() => {
    setValue(props.phrase);
  }, [props.phrase]);

  const dispatch = useDispatch();

  function handleSubmit() {
    if (value === undefined) {
      return;
    }
    EditSearchPhrase(value).then((res) => {
      if (props.onClose) {
        props.onClose();
        setError('');
      }

      dispatch(fetchPhrases());
    }).catch((err) => {
      setError(err.response.data);
    });
  }

  return <Overlay open={props.open} onExit={() => {
    if (props.onClose) {
      props.onClose();
    }
    setError('');
  }}>
    Edit a Search Phrase
    {error && <ErrorText>{error}</ErrorText>}
    {value !== undefined && <SearchPhraseForm value={value} onValueChange={setValue} onSubmit={handleSubmit}/>}
  </Overlay>;
}
