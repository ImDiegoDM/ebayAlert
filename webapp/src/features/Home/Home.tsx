import { faPlus } from '@fortawesome/free-solid-svg-icons';
import { FontAwesomeIcon } from '@fortawesome/react-fontawesome';
import Axios from 'axios';
import React, { useEffect, useState } from 'react';
import { useDispatch } from 'react-redux';
import styled from 'styled-components';
import { fetchPhrases } from '../../actions';
import { modalOpen } from '../../actions/modal';
import { ISearchPhrase } from '../../interfaces';
import { Header } from './components/Header';
import { Phrases } from './components/Phrases';
import { SearchBar } from './components/SearchBar';
import { CreateOverlay } from './CreateOverlay';
import { EditOverlay } from './EditOverlay';

type OverlayType = 'create' | 'edit' | 'closed';

const Add = styled.div`
  color:gray;
  cursor: pointer;
`;

export function Home() {
  const [overlay, setOverlay] = useState<OverlayType>('closed');
  const [selectedPhrase, setSelectedPhrase] = useState<ISearchPhrase|undefined>(undefined);
  const dispatch = useDispatch();

  const [search, setSearch] = useState('');

  const [timeout, settime] = useState<any>(undefined);

  useEffect(() => {
    clearTimeout(timeout);

    settime(setTimeout(() => {
      dispatch(fetchPhrases(search));
    }, 300));

  }, [search]);

  function handleDelete(id?: string) {
    if (id) {
      dispatch(modalOpen(async () => {
        await Axios.delete(process.env.REACT_APP_SERVER_URL + '/search_phrases/' + id);
        dispatch(fetchPhrases());
      }, 'Are you sure that you want to delete this alert ?'));
    }
  }

  return <>
    <Header>
      <div>Ebay Alert Program</div>
      <SearchBar value={search} onChange={setSearch}/>
      <Add onClick={() => setOverlay('create')}>
        <FontAwesomeIcon icon={faPlus}/>
      </Add>
    </Header>
    <Phrases onDelete={handleDelete} onSelect={(phrase) => {
      setSelectedPhrase(phrase);
      setOverlay('edit');
    }}/>
    <CreateOverlay open={overlay === 'create'} onClose={() => setOverlay('closed')}/>
    <EditOverlay
      phrase={selectedPhrase}
      open={overlay === 'edit' && selectedPhrase !== undefined}
      onClose={() => setOverlay('closed')}/>
  </>;
}
