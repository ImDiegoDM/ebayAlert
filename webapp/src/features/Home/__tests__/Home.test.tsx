import { render } from '@testing-library/react';
import React from 'react';
import { Provider } from 'react-redux';
import { applyMiddleware, createStore } from 'redux';
import promise from 'redux-promise-middleware';
import { ISearchPhrase } from '../../../interfaces';
import reducer from '../../../reducers';
import { Home } from '../Home';

const phrase: ISearchPhrase = {
  _id: 'sdfd5fa1dfad',
  email: 'test@test.com',
  howOften: '2',
  phrase: 'drone',
};

const searchPhrases = {
  data: [phrase],
  fetched: true,
  fetching: false,
  ok: true,
};

const searchPhrasesFetching = {
  data: undefined,
  fetched: false,
  fetching: true,
  ok: false,
};

describe('<Home/>', () => {
  it('should match snapshot when store state is fetched', () => {
    const middlware = applyMiddleware(promise);

    const store = createStore(reducer, {searchPhrases}, middlware);

    const { container } = render(
      <Provider store={store}>
        <Home />
      </Provider>,
    );

    expect(container.firstChild).toMatchSnapshot();
  });

  it('should match snapshot when store state is fetching', () => {
    const middlware = applyMiddleware(promise);

    const store = createStore(reducer, {
      searchPhrases: searchPhrasesFetching,
    }, middlware);

    const { container } = render(
      <Provider store={store}>
        <Home />
      </Provider>,
    );

    expect(container.firstChild).toMatchSnapshot();
  });
});
