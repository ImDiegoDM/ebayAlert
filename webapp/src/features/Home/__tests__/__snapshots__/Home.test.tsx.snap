// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Home/> should match snapshot when store state is fetched 1`] = `
.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  width: 100%;
  padding: 15px;
  border-bottom: 2px solid #ececec;
  box-sizing: border-box;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: justify;
  -webkit-justify-content: space-between;
  -ms-flex-pack: justify;
  justify-content: space-between;
}

.c1 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  border: 1px solid #b7b5b5;
  border-radius: 5px;
  padding: 5px 10px;
}

.c2 {
  color: gray;
  margin-right: 5px;
}

.c3 {
  border: none;
}

.c3:focus {
  outline: none;
}

.c4 {
  color: gray;
  cursor: pointer;
}

<div
  class="c0"
>
  <div>
    Ebay Alert Program
  </div>
  <div
    class="c1"
  >
    <svg
      aria-hidden="true"
      class="svg-inline--fa fa-search fa-w-16 c2"
      data-icon="search"
      data-prefix="fas"
      focusable="false"
      role="img"
      viewBox="0 0 512 512"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        d="M505 442.7L405.3 343c-4.5-4.5-10.6-7-17-7H372c27.6-35.3 44-79.7 44-128C416 93.1 322.9 0 208 0S0 93.1 0 208s93.1 208 208 208c48.3 0 92.7-16.4 128-44v16.3c0 6.4 2.5 12.5 7 17l99.7 99.7c9.4 9.4 24.6 9.4 33.9 0l28.3-28.3c9.4-9.4 9.4-24.6.1-34zM208 336c-70.7 0-128-57.2-128-128 0-70.7 57.2-128 128-128 70.7 0 128 57.2 128 128 0 70.7-57.2 128-128 128z"
        fill="currentColor"
      />
    </svg>
    <input
      class="c3"
      type="text"
      value=""
    />
  </div>
  <div
    class="c4"
  >
    <svg
      aria-hidden="true"
      class="svg-inline--fa fa-plus fa-w-14 "
      data-icon="plus"
      data-prefix="fas"
      focusable="false"
      role="img"
      viewBox="0 0 448 512"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        d="M416 208H272V64c0-17.67-14.33-32-32-32h-32c-17.67 0-32 14.33-32 32v144H32c-17.67 0-32 14.33-32 32v32c0 17.67 14.33 32 32 32h144v144c0 17.67 14.33 32 32 32h32c17.67 0 32-14.33 32-32V304h144c17.67 0 32-14.33 32-32v-32c0-17.67-14.33-32-32-32z"
        fill="currentColor"
      />
    </svg>
  </div>
</div>
`;

exports[`<Home/> should match snapshot when store state is fetching 1`] = `
.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  width: 100%;
  padding: 15px;
  border-bottom: 2px solid #ececec;
  box-sizing: border-box;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: justify;
  -webkit-justify-content: space-between;
  -ms-flex-pack: justify;
  justify-content: space-between;
}

.c1 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  border: 1px solid #b7b5b5;
  border-radius: 5px;
  padding: 5px 10px;
}

.c2 {
  color: gray;
  margin-right: 5px;
}

.c3 {
  border: none;
}

.c3:focus {
  outline: none;
}

.c4 {
  color: gray;
  cursor: pointer;
}

<div
  class="c0"
>
  <div>
    Ebay Alert Program
  </div>
  <div
    class="c1"
  >
    <svg
      aria-hidden="true"
      class="svg-inline--fa fa-search fa-w-16 c2"
      data-icon="search"
      data-prefix="fas"
      focusable="false"
      role="img"
      viewBox="0 0 512 512"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        d="M505 442.7L405.3 343c-4.5-4.5-10.6-7-17-7H372c27.6-35.3 44-79.7 44-128C416 93.1 322.9 0 208 0S0 93.1 0 208s93.1 208 208 208c48.3 0 92.7-16.4 128-44v16.3c0 6.4 2.5 12.5 7 17l99.7 99.7c9.4 9.4 24.6 9.4 33.9 0l28.3-28.3c9.4-9.4 9.4-24.6.1-34zM208 336c-70.7 0-128-57.2-128-128 0-70.7 57.2-128 128-128 70.7 0 128 57.2 128 128 0 70.7-57.2 128-128 128z"
        fill="currentColor"
      />
    </svg>
    <input
      class="c3"
      type="text"
      value=""
    />
  </div>
  <div
    class="c4"
  >
    <svg
      aria-hidden="true"
      class="svg-inline--fa fa-plus fa-w-14 "
      data-icon="plus"
      data-prefix="fas"
      focusable="false"
      role="img"
      viewBox="0 0 448 512"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        d="M416 208H272V64c0-17.67-14.33-32-32-32h-32c-17.67 0-32 14.33-32 32v144H32c-17.67 0-32 14.33-32 32v32c0 17.67 14.33 32 32 32h144v144c0 17.67 14.33 32 32 32h32c17.67 0 32-14.33 32-32V304h144c17.67 0 32-14.33 32-32v-32c0-17.67-14.33-32-32-32z"
        fill="currentColor"
      />
    </svg>
  </div>
</div>
`;
